File: 二分查找/50.py
'''
实现 pow(x, n) ，即计算 x 的 n 次幂函数，即x^n）。
'''


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1
        if n % 2 == 0:
            if n > 0:
                return self.myPow(x, n/2)*self.myPow(x, n/2)
            else:
                return 1/(self.myPow(x, -n/2)*self.myPow(x, -n/2))
        if n % 2 == 1:
            if n > 0:
                return self.myPow(x, n-1)*x
            else:
                return 1/(self.myPow(x, -n-1)*x)

    def _myPow(self, x: float, n: int) -> float:
        ans = 1
        abs_n = abs(n)
        while(abs_n):
            if abs_n%2==1:
                ans = ans * x
            x = x*x
            abs_n = abs_n//2
        return ans if n>=0 else 1/ans

x = 2
n = -4
obj = Solution()
print(obj._myPow(x, n))

File: 二分查找/50.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    //迭代快速幂 O(logn),O(logn)
    double qPow(double x, long long N)
    {
        if (N == 0)
            return 1;
        double temp = qPow(x, N / 2);
        return N % 2 == 0 ? temp * temp : temp * temp * x;
    }
    //非迭代快速幂 O(logn),O(1)
    double loopPow(double x, long long N)
    {
        double res = 1;
        while (N)
        {
            if (N % 2 == 1)
                res = res * x;
            x = x * x;
            N = N / 2;
        }
        return res;
    }
    double myPow(double x, int n)
    {
        long long N = n;
        return N >= 0 ? loopPow(x, N) : 1 / loopPow(x, -N);
    }
};

int main()
{
    double x = 2;
    int n = -3;
    Solution obj;
    cout << obj.myPow(x, n) << endl;
}